boa/contracts/abi/contract.py

```diff
@@ -0,0 +1,120 @@
+from functools import cached_property
+from operator import attrgetter
+from os.path import basename
+from typing import Any, Optional, Union
+from warnings import warn
+
+from eth.abc import ComputationAPI
+
+from boa.contracts.abi import _decode_addresses, _format_abi_type, group_by
+from boa.contracts.abi.function import ABIFunction, ABIOverload
+from boa.contracts.evm_contract import BaseEVMContract
+from boa.environment import Address
+from boa.util.abi import abi_decode
+from boa.util.exceptions import StackTrace, _handle_child_trace
+
+
+class ABIContract(BaseEVMContract):
+    """A contract that has been deployed to the blockchain and created via an ABI."""
+
+    def __init__(
+        self,
+        name: str,
+        functions: list[ABIFunction],
+        address: Address,
+        filename: Optional[str] = None,
+        env=None,
+    ):
+        super().__init__(env, filename=filename, address=address)
+        self._name = name
+        self._functions = functions
+        self._bytecode = self.env.vm.state.get_code(address.canonical_address)
+        if not self._bytecode:
+            warn(
+                f"Requested {self} but there is no bytecode at that address!",
+                stacklevel=2,
+            )
+
+        for name, group in group_by(self._functions, attrgetter("name")).items():
+            setattr(self, name, ABIOverload.create(group, self))
+
+        self._address = Address(address)
+
+    @cached_property
+    def method_id_map(self):
+        """
+        Returns a mapping from method id to function object.
+        This is used to create the stack trace when an error occurs.
+        """
+        return {function.method_id: function for function in self._functions}
+
+    def marshal_to_python(self, computation, abi_type: list[str]) -> tuple[Any, ...]:
+        """
+        Convert the output of a contract call to a Python object.
+        :param computation: the computation object returned by `execute_code`
+        :param abi_type: the ABI type of the return value.
+        """
+        if computation.is_error or (abi_type and not computation.output):
+            return self.handle_error(computation)
+
+        schema = f"({_format_abi_type(abi_type)})"
+        decoded = abi_decode(schema, computation.output)
+        return tuple(_decode_addresses(typ, val) for typ, val in zip(abi_type, decoded))
+
+    def stack_trace(self, computation: ComputationAPI) -> StackTrace:
+        """
+        Create a stack trace for a failed contract call.
+        """
+        calldata_method_id = bytes(computation.msg.data[:4])
+        if calldata_method_id in self.method_id_map:
+            function = self.method_id_map[calldata_method_id]
+            msg = f"  (unknown location in {self}.{function.full_signature})"
```

Comment:
i don't think we need the `unknown location` -- it's pretty clear from the fact that it's an ABI contract.